### PR TITLE
nomic-embed-text model implementation 

### PR DESCRIPTION
**Pull request by npardal**
This PR introduces the nomic-embed-text model implementation for v1.5

### CODE REVIEW

File: model/models/nomicbert/model.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+	hiddenStates := m.TokenEmbedding.Forward(ctx, batch.Inputs)
+
+	typeEmbed := m.TypeEmbedding.Weight.View(ctx, 0, m.hiddenSize)
+	hiddenStates = hiddenStates.Add(ctx, typeEmbed)
+
+	// no position embedding - we use ROPE instead
+
+	hiddenStates = m.TokenEmbeddingNorm.Forward(ctx, hiddenStates, m.eps)
+
+	positions := ctx.Arange(0, float32(seqLen), 1, ml.DTypeI32)
```
Comment (mxyng) [inline model/models/nomicbert/model.go]: positional ids are already set by the caller

```suggestion
	positions := ctx.Input().FromInts(batch.Positions, len(batch.Positions))
```

File: model/models/nomicbert/model.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+	Down *nn.Linear `gguf:"ffn_down"`
+}
+
+// Forward implements model.Model interface
+func (m *Model) Forward(ctx ml.Context, batch input.Batch) (ml.Tensor, error) {
+	seqLen := batch.Inputs.Dim(0)
+
+	hiddenStates := m.TokenEmbedding.Forward(ctx, batch.Inputs)
+
+	typeEmbed := m.TypeEmbedding.Weight.View(ctx, 0, m.hiddenSize)
```
Comment (mxyng) [inline model/models/nomicbert/model.go]: If you rebase, you'll be able to use the slice operation


```suggestion
	typeEmbed := m.TypeEmbedding.Weight.Slice(ctx, 1, 0, 1, 1)
```

File: model/models/nomicbert/model.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+	return hiddenStates
+}
+
+func (a *Attention) Forward(ctx ml.Context, hiddenStates ml.Tensor, positions ml.Tensor, opts *Options) ml.Tensor {
+	batchSize := hiddenStates.Dim(1)
+
+	qkv := a.QKV.Forward(ctx, hiddenStates)
+
+	query := qkv.View(ctx, 0, opts.hiddenSize, qkv.Stride(0), batchSize).Contiguous(ctx)
```
Comment (mxyng) [inline model/models/nomicbert/model.go]: This can also be replaced with `tensor.Slice(...)`

File: model/models/nomicbert/model.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+	attention = attention.Reshape(ctx, opts.hiddenSize, batchSize)
+
+	return a.Output.Forward(ctx, attention)
+}
+
+func (m *MLP) Forward(ctx ml.Context, hiddenStates ml.Tensor) ml.Tensor {
+	gate := m.Gate.Forward(ctx, hiddenStates).SILU(ctx)
+	up := m.Up.Forward(ctx, hiddenStates)
+	hidden := gate.Mul(ctx, up)
```
Comment (mxyng) [inline model/models/nomicbert/model.go]:
```suggestion
	hidden := m.Gate.Forward(ctx, hiddenStates).SILU(ctx, m.Up.Forward(ctx, hiddenStates))
```

File: model/models/nomicbert/model.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+			eps:          c.Float("attention.layer_norm_epsilon"),
+			poolingType:  pooling.Type(c.Uint("pooling_type")),
+			normalize:    c.Bool("normalize_embeddings", true),
+			ropeFreqBase: c.Float("rope.freq_base", 1000.0),
+		},
+	}, nil
+}
+
+func init() {
+	model.Register("nomic-bert", New)
```
Comment (mxyng) [inline model/models/nomicbert/model.go:168]: you'll need to import this package in `model/models/models.go` in order to fully register the model

File: model/models/nomicbert/model.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+	return m.Down.Forward(ctx, hidden)
+}
+
+func New(c fs.Config) (model.Model, error) {
+	hiddenSize := int(c.Uint("embedding_length"))
+	numHeads := int(c.Uint("attention.head_count"))
+	headDim := hiddenSize / numHeads
+
+	processor := model.NewWordPiece(
```
Comment (mxyng) [inline model/models/nomicbert/model.go:129]: this model doesn't want its inputs lowercased. there's a todo in `WordPiece.Encode(...)` to handle this case. you should make that change as well

File: model/models/nomicbert/model.go [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+	return &Model{
+		TextProcessor: processor,
+		Layers:        make([]EncoderLayer, c.Uint("block_count")),
+		Options: Options{
+			hiddenSize:   hiddenSize,
+			numHeads:     numHeads,
+			headDim:      headDim,
+			eps:          c.Float("attention.layer_norm_epsilon"),
+			poolingType:  pooling.Type(c.Uint("pooling_type")),
+			normalize:    c.Bool("normalize_embeddings", true),
```
Comment (mxyng) [inline model/models/nomicbert/model.go]: the models doesn't want its embeddings normalized so this should default to false

File: model/models/nomicbert/model.go [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+	qkv := a.QKV.Forward(ctx, hiddenStates)
+
+	query := qkv.Slice(ctx, 0, 0, opts.hiddenSize, 1).Contiguous(ctx)
+	key := qkv.Slice(ctx, 0, opts.hiddenSize, 2*opts.hiddenSize, 1).Contiguous(ctx)
+	value := qkv.Slice(ctx, 0, 2*opts.hiddenSize, 3*opts.hiddenSize, 1).Contiguous(ctx)
+
+	query = query.Reshape(ctx, opts.headDim, opts.numHeads, batchSize)
+	key = key.Reshape(ctx, opts.headDim, opts.numHeads, batchSize)
+	value = value.Reshape(ctx, opts.headDim, opts.numHeads, batchSize)
```
Comment (mxyng) [inline model/models/nomicbert/model.go]: You should reshape before slice which cleans this up significantly and allows skipping the contiguous calls

```go
qkv = qkv.Reshape(ctx, opts.headDim, -1, batchSize)
chunks = qkv.Chunk(ctx, 1, opts.numHeads)
query, key, value = chunks[0], chunks[1], chunks[2]
```
